### PR TITLE
[docs] 2/n add airflow migration strategies: TaskFlow, Kubernetes

### PR DESCRIPTION
**Pull request by cmpadden**
## Summary & Motivation

- Removes `dagster-airflow` migration strategy
- Introduces "Choose your own adventure" migrations
    * Pipelines using Airflow TaskFlow API
    * Pipelines using Airflow KubernetesPodOperator

### CODE REVIEW

**Comment by cmpadden** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22614?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#22614** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22614?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#22612** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22612?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @cmpadden and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 1]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
 
 # Migrating Airflow to Dagster
 
-<Note>
-  Looking for an example of an Airflow to Dagster migration? Check out the{" "}
-  <a href="https://github.com/dagster-io/dagster-airflow-migration-example">
```
Comment (cmpadden) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:10]: Reminder to archive this repository as we deprecate `dagster-airflow`

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 2]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
...
-### Containerized operator considerations
+```python
+import json
 
-There are a variety of Airflow Operator types that are used to launch compute in various external execution environments, for example Kubernetes or Amazon ECS. When getting things working locally we'd recommend trying to execute those containers locally unless it's either unrealistic or impossible to emulate the cloud environment. For example if you use the K8sPodOperator, it likely means that you will need to have a local Kubernetes cluster running, and in that case we recommend docker's built-in Kubernetes environment. You also need to be able to pull down the container images that will be needed for execution to your local machine.
+from dagster import Definitions, MaterializeResult, define_asset_job, asset
 
-If local execution is impossible, we recommend using Branch Deployments in Dagster+, which is a well-supported workflow for cloud-native development.
+@asset
+def extract_transform_load_asset():
```
Comment (cmpadden) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:150]: Would like feedback as to whether or not we want to show both the `@op` approach and the consolidated asset.

Originally I had done three assets: extract, transform, and load -- would that be preferred here?

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 3]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
 
 # Migrating Airflow to Dagster
```
Comment (cmpadden) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:6]: How do we feel about the structure of this page - do we need an index at the top? Should it be split into multiple pages?

We will continue to append new migration strategies / mapping of concepts over time.

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx (hunk 2 above)
Comment (maximearmstrong) [reply docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:150]: I would choose any asset variation over the standalone `@op` approach. Using `@op` in a graph-backed asset could be a nice middle ground. 

Having three assets also makes sense, as users may want a MaterializeResult for each of them.

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 4]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
...
+                    "name": "data-processing-rs",
+                    "image": "data-processing-rs",
+                }
+            ]
+        },
+    ).get_materialize_result()
+```
 
----
+The primary difference between Airflow and Dagster are how the k8s pod specification are expose. In Airflow, they are passed as parameters the `KubernetesPodOperator`, whereas in Dagster they are passed as a `base_pod_spec` dictionary to the `k8s_pipes_client.run` method. Another difference is that in Airflow, `get_logs` is required to capture _stdout_, however, with Dagster they are automatically captured to on the _stdout_ tab of the step output.
```
Comment (erinkcochran87) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:233]:
```suggestion
The primary difference between Airflow and Dagster are how the k8s pod specifications are exposed. In Airflow, they are passed as parameters to the `KubernetesPodOperator`, whereas in Dagster they are passed as a `base_pod_spec` dictionary to the `k8s_pipes_client.run` method. Additionally, in Airflow, `get_logs` is required to capture `stdout`. In Dagster, however, they are automatically captured on the `stdout` tab of the step output.
```

For the `stdout` tab, are you referring to the tab in the Run Details page of the UI?

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 5]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
...
 
-## Step 8: Migrate permissions to Dagster
+**Advanced pipes usage**
 
-Your Airflow instance likely had specific IAM or Kubernetes permissions that allowed it to successfully run your Airflow DAGs. To run the migrated Dagster jobs, you'll need to duplicate these permissions for Dagster.
+In the above example, we demonstrated how to easily run images on Kubernetes using the `dagsater_k8s` library. One of the biggest benefits of pipes, that we've not yet covered, is that you can leverage the `dagster_pipes` library from within your containerized pipeline to access the full Dagster context, and emit events back to the Dagster UI!
```
Comment (erinkcochran87) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:237]:
```suggestion
In the above example, we demonstrated how to run images on Kubernetes using the `dagsater_k8s` library. One of the biggest benefits of Dagster Pipes, however, is that you can leverage the `dagster_pipes` library from within your containerized pipeline to access the full Dagster context, and emit events back to the Dagster UI.
```

Some suggestions for improving readability

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 6]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
...
 
-- **We recommend using [Airflow connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) or [environment variables](/dagster-plus/managing-deployments/environment-variables-and-secrets)** to define permissions whenever possible.
+A common pattern when building containerized pipelines is to accept a large number of command-line arguments using libraries like `argparse`. However, with Dagster it is possible pass a dictionary of parameters on the Dagster context using the `extras` parameter. Then, in your pipeline code you can access the context `PipesContext.get()` if you are using Python.
```
Comment (erinkcochran87) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:239]:
```suggestion
A common pattern when building containerized pipelines is to accept a large number of command-line arguments using libraries like `argparse`. However, with Dagster you can pass a dictionary of parameters on the Dagster context using the `extras` parameter. Then, in your pipeline code, you can access the context `PipesContext.get()` if you are using Python.
```

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 7]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
...
 
-- **If you're unable to use Airflow connections or environment variables,** you can attach permissions directly to the infrastructure where you're deploying Dagster.
+<Note>
+  Support for Pipes clients in other programming languages are in the works, if
+  you have any specific requirements, please let us know!
+</Note>
 
-- **If your Airflow DAGs used [`KubernetesPodOperators`](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html)**, it's possible that you loaded a `kube_config` file or used the `in_cluster` config. When migrating, we recommend switching to [using connections with a `kube_config` JSON blob](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/connections/kubernetes.html) to make things easier.
+For a full tutorial on how to use Dagster Pipes, please reference [Dagster Pipes tutorial](https://docs.dagster.io/concepts/dagster-pipes/subprocess).
```
Comment (erinkcochran87) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:246]:
```suggestion
For a step-by-step walkthrough of using Dagster Pipes, refer to the [Dagster Pipes tutorial](https://docs.dagster.io/concepts/dagster-pipes/subprocess).
```

Kinda nitpicky, but using `please` in tech docs isn't recommended

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 8]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
...
 
-- **We recommend using [Airflow connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) or [environment variables](/dagster-plus/managing-deployments/environment-variables-and-secrets)** to define permissions whenever possible.
+A common pattern when building containerized pipelines is to accept a large number of command-line arguments using libraries like `argparse`. However, with Dagster it is possible pass a dictionary of parameters on the Dagster context using the `extras` parameter. Then, in your pipeline code you can access the context `PipesContext.get()` if you are using Python.
 
-- **If you're unable to use Airflow connections or environment variables,** you can attach permissions directly to the infrastructure where you're deploying Dagster.
+<Note>
+  Support for Pipes clients in other programming languages are in the works, if
+  you have any specific requirements, please let us know!
+</Note>
```
Comment (erinkcochran87) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:244]: Why include this here?

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 9]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
...
+            ]
+        },
+    ).get_materialize_result()
+```
 
----
+The primary difference between Airflow and Dagster are how the k8s pod specification are expose. In Airflow, they are passed as parameters the `KubernetesPodOperator`, whereas in Dagster they are passed as a `base_pod_spec` dictionary to the `k8s_pipes_client.run` method. Another difference is that in Airflow, `get_logs` is required to capture _stdout_, however, with Dagster they are automatically captured to on the _stdout_ tab of the step output.
 
-## Step 8: Migrate permissions to Dagster
+**Advanced pipes usage**
```
Comment (erinkcochran87) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:235]: This should be a subheading, but I'm not sure what level is should be at the moment. We can fix it once we figure out what the structure of this page is going to look like.

**Review by erinkcochran87 — COMMENTED** [review]
Back to you!

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx (hunk 8 above)
Comment (cmpadden) [reply docs/content/guides/migrations/migrating-airflow-to-dagster.mdx]: Removed

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx (hunk 4 above)
Comment (cmpadden) [reply docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:235]: Yes, that's right.

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx (hunk 2 above)
Comment (cmpadden) [reply docs/content/guides/migrations/migrating-airflow-to-dagster.mdx]: For now I've removed the `@op` example and replaced it with a 1-to-1 mapping of `@asset`. This is using the implicit asset outs to keep it simple, as I agree that introducing graph-backed right away may be overwhelming.

Comment (yuhan) [reply docs/content/guides/migrations/migrating-airflow-to-dagster.mdx]: Let's focus on asset!

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx (hunk 1 above)
Comment (cmpadden) [reply docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:10]: Repository has been archived.

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 10]
```diff
@@ -5,223 +5,242 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
...
-- If you rely on the usage of the Airflow UI to set production connections or variables, this will require a change of workflow as you will no longer have the Airflow UI at the end of this migration. If instead you rely on code-as-infrastructure patterns to set connections or variables, migration is more straightforward.
-- If you have standardized on a small set or even a single operator type (e.g. the `K8sPodOperator`), migration will be easier. If you use a large number of operator types with a wide range of infrastructure requirements, migration will be more work.
-- If you dynamically generate your Airflow DAGs from a higher-level API or DSL (e.g. yaml), the migration will be more straightforward than all your stakeholders directly creating Airflow DAGs.
+@dag(
+    schedule=None,
+    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
+    catchup=False,
+    tags=["example"],
+)
+def tutorial_taskflow_api():
```
[the remark this answers is not in the document]
Comment (cmpadden) [reply docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:59]: Ok, I've removed the `@op` example and added smaller snippets of the task and asset before showing the full examples.

File: docs/content/guides/migrations/migrating-airflow-to-dagster.mdx [hunk 11]
```diff
@@ -5,223 +5,229 @@ description: "Learn how do a lift-and-shift migration of airflow to Dagster."
...
 
-Your Airflow instance likely had specific IAM or Kubernetes permissions that allowed it to successfully run your Airflow DAGs. To run the migrated Dagster jobs, you'll need to duplicate these permissions for Dagster.
+### Advanced pipes usage
 
-- **We recommend using [Airflow connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) or [environment variables](/dagster-plus/managing-deployments/environment-variables-and-secrets)** to define permissions whenever possible.
+In the above example, we demonstrated how to run images on Kubernetes using the `dagsater_k8s` library. One of the biggest benefits of Dagster Pipes, however, is that you can leverage the `dagster_pipes` library from within your containerized pipeline to access the full Dagster context, and emit events back to the Dagster UI.
```
Comment (maximearmstrong) [inline docs/content/guides/migrations/migrating-airflow-to-dagster.mdx:231]:
```suggestion
In the above example, we demonstrated how to run images on Kubernetes using the `dagster_k8s` library. One of the biggest benefits of Dagster Pipes, however, is that you can leverage the `dagster_pipes` library from within your containerized pipeline to access the full Dagster context, and emit events back to the Dagster UI.
```

**Review by maximearmstrong — APPROVED** [review]
Only one typo left, but other than that it's great 💯

**Comment by cmpadden** [comment]
### Merge activity

* **Jun 28, 4:14 PM EDT**: @cmpadden started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/22614).

**Pull request merged**